README
README file for Pgtcl/libpgtcl

$Id: README,v 1.13 2005/04/20 19:54:03 schwarzkopf Exp $

Thank you for downloading Pgtcl, a package that adds PostgreSQL interface
extensions to the Tcl programming language... an open source project that's 
been in existence for more than ten years.

CONFIGURING

Pgtcl is now Tcl Extension Architecture (TEA) compliant, shipping with
a standard "configure" script.  It no longer needs to reside in a specific
place within the Postgres source tree in order to build.  

For standard builds to put things in subdirectories of /usr/local, you
can often simply execute configure with no arguments at all...

./configure

The configure script will attempt to find where the Postgres includes and
libraries are using pg_config, a program built and installed as part of
Postgres.  Alternatively, you can specify a path to the Postgres 
include files using --with-postgres-include and to the Postgres libraries 
using --with-postgres-lib.  

If you had PostgreSQL installed into /usr/postgres and a Tcl build in
/usr/pptcl, you might use something like

./configure --prefix=/usr/pptcl

With this style of configure command, you'll need to make sure pg_config
(and the correct pg_config if you have postgres installed in multiple
places) is in the PATH.

Alternatively you can still explicitly specify where the Postgres includes
and libraries are found:

./configure --prefix=/usr/pptcl --with-postgres-include=/usr/postgres/include --with-postgres-lib=/usr/postgres/lib

The other configure parameters that may need tweaking are where Tcl's
includes and libraries (and tclConfig.sh) are.  Although normally they
will be in /usr/local/include and /usr/local/lib, in some cases they
may reside elsewhere.  If Tcl is built and installed from the FreeBSD ports 
tree, for example, they'll be in versioned subdirectories...

./configure  --with-tcl=/usr/local/lib/tcl8.4 --with-tclinclude=/usr/local/inclu
de/tcl8.4


BUILDING

Do a "make".  If all goes well, libpgtcl will be compiled and linked.

INSTALLING

Do a "make install".

USING IT

With version 1.4, Pgtcl is a standard package and can be loaded with
"package require" instead of the shared library load routine, "load".

Fire up your tclsh:

tclsh8.4
% package require Pgtcl
1.4

It's a good idea to switch to using the "package require' instead of "load"
to pick up Pgtcl, because there will be additional Tcl code shipped in future 
versions of Pgtcl, and using "package require" will make that code available 
to your application.  Also it keeps you from hard-coding the path to the
library and hard-coding a dependency on a specific version.

IF IT COMPILES AND INSTALLS OK BUT "PACKAGE REQUIRE" DOESN'T WORK

...it probably didn't install into the search path Tcl uses to find
extensions.  You might have more than one Tclsh installed.  Try adding
a path to the parent directory of where the thing installed, for example,
if it installed into /opt/local/bin

tclsh8.4
% lappend auto_path /opt/local/lib
...
% package require Pgtcl
1.4

CREDITS

Pgtcl was originally written by Jolly Chen.  Many people have contributed to 
the further development of Pgtcl over the years, including Randy Kunkee,
who added the channel handler code, among other things, and we intend to
identify the rest of them and give them credit as well.

Development and maintenance of Pgtcl since version 1.3 has been done by Brett 
Schwarz and Karl Lehenbauer.

VERSION 1.5

New options to pg_connect, -connhandle and -connlist

experimental -dict option in pg_result

Proper Tcl namespace support, in the ::pg namespace.

Connection handles and result handles are now executable commands in their
own right, while maintaining compatible with their former usage as strictly
handles.

Upgraded to TEA 3.1 compliant build.

Documentation overhauled and brought current with the code.

VERSION 1.4

With version 1.4, Pgtcl has been internally overhauled and brought up to
date with the latest Tcl C-interface technology, while maintaining
nearly 100% compatibility with the pg_* Tcl interface provided by Pgtcl 1.3.

Just about every Tcl program that uses Pgtcl 1.3 will work without modification
under Pgtcl 1.4.

Version 1.4 was something of a transitional release, as pgtcl moved out the 
of core and into its this distribution.  Previously, for example, the
Pgtcl documentation resided with the rest of the PostgreSQL documentation,
and Pgtcl's source code accessed PostgreSQL's include files and libraries
in a fraternal manner that had to be divorced and reworked to use the same
APIs and build methods that any external application would use to build against
PostgreSQL's libpq C interface library.  

The Pgtcl documentation is now included with this release.  As building the
documentation requires a number of fairly major tools and packages, the
release also includes the docs prebuilt in HTML and PDF format.

CHANGES

The main changes are:

    o All commands have now been converted to use Tcl 8-style Tcl objects.

        The result is a probable increase in performance in all routines, with
        potentially huge performance increases in pg_select and pg_execute when
        used with complex Tcl code bodies.

	Also Tcl 7 is no longer supported.  (Surely you're not still
	using it, anyway, right?)

    o A new asynchronous interface has been added

	Requests can be issued to the backend without waiting for the
	results, allowing for user interfaces to still work while
	database requests are being processed, etc.
	Also, requests can now be cancelled while they're in progress.

    o pg_* call arguments are now checked much more rigorously.

	Code previously using atoi() for integer conversions now
	uses Tcl_GetIntFromObj, etc.

	pg_* calls with too many arguments were often accepted without
	complaint.  These now generate standard Tcl "wrong # args"
	messages, etc.

    o Error reporting has been brought into more compliance with the
      Tcl way of doing things.

    o TEA-compliant build and install.

    o pg_exec now supports $-variable substitution.  This is a new
      feature of Postgres 7.4 that is now supported in Pgtcl.
      You can now say

	pg_exec $conn {select * from foo where id = $1 and user = $2} $id $user

    And the values following the SQL statement are substituted positionally
    for $1, $2, etc, in the statement.

    o pg_exec_prepared allows execution of prepared SQL statements.

    o pg_sendquery_prepared allows asynchronous execution of prepared
      SQL statements.


Some programs that might have been working properly but had certain
syntatically incorrect pg_* commands will now fail until fixed.

pg_result -assign and pg_result -assignbyidx used to return the array
name, which was superfluous because the array name was specified on the
command line.  They now return nothing.  *** POTENTIAL INCOMPATIBILITY ***

OLDER CHANGES

Here are some features that were added back in the 1998 - 1999 timeframe:

1. Postgres connections are a valid Tcl channel, and can therefore
   be manipulated by the interp command (ie. shared or transfered).
   A connection handle's results are transfered/shared with it.
   (Result handles are NOT channels, though it was tempting).  Note
   that a "close $connection" is now functionally identical to a
   "pg_disconnect $connection", although pg_connect must be used
   to create a connection.
   
2. Result handles are changed in format: ${connection}.<result#>.
   This just means for a connection 'pgtcl0', they look like pgtcl0.0,
   pgtcl0.1, etc.  Enforcing this syntax makes it easy to look up
   the real pointer by indexing into an array associated with the
   connection.

3. I/O routines are now defined for the connection handle.  I/O to/from
   the connection is only valid under certain circumstances: following
   the execution of the queries "copy <table> from stdin" or
   "copy <table> to stdout".  In these cases, the result handle obtains
   an intermediate status of "PGRES_COPY_IN" or "PGRES_COPY_OUT".  The
   programmer is then expected to use Tcl gets or read commands on the
   database connection (not the result handle) to extract the copy data.
   For copy outs, read until the standard EOF indication is encountered.
   For copy ins, puts a single terminator (\.).  The statement for this
   would be
	puts $conn "\\."      or       puts $conn {\.}
   In either case (upon detecting the EOF or putting the `\.', the status
   of the result handle will change to "PGRES_COMMAND_OK", and any further
   I/O attempts will cause a Tcl error.




libpgtcl-1.7/ChangeLog

$Id: ChangeLog,v 1.57 2009/04/06 15:22:01 karl Exp $

2009-04-06 Karl Lehenbauer
    * Pgtcl 1.7 released.

2009-02-26 Karl Lehenbauer
    * Make PostgreSQL helpers "res_dont_care" and "res_must_succeed"
      accept PGRES_TUPLES_OK as a sucessful result as well as
      PGRES_COMMAND_OK.

2008-05-24 Karl Lehenbauer
    * Add new pg_result suboption -tupleArrayWithoutNulls.  It works like
      -tupleArray except if a field is null, it unsets the field from
      the array.

2008-05-23 Karl Lehenbauer
    * Add "escape_bytea" and "unescape_bytea" to the methods that can be
      invoked from $conn.

    * Extend pg_escape_bytea to allow a connection handle to be specified.
      If specified, libpq's PQescapeByteaConn is used rather than
      deprecated PQescapeBytea, which can return incorrect results since
      it can't determine whether standard-conforming strings are enabled.

      If the "$conn escape_bytea ..." method is used, the connection-handle
      version is automatically used.

    * Detect if lo_truncate is available in libpq or not and, if not, return 
      an error if it is attempted.

2008-05-20 Karl Lehenbauer
    * Add support for the lo_truncate function that truncates a large object
      to a given length.

    * Version bump to 1.7.

2007-12-14 Karl Lehenbauer <karl-pg@sc.com>
    * Update autoconf build to Tcl Extension Architecture (TEA) 3.6

    * Extend pg_quote to allow a connection handle to be specified.
      If the connection handle is specified and the string matches the 
      null string, returns "NULL" (without quotes).  If not, returns 
      the string quoted for inclusion in SQL commands.

    * pg_quote can be invoked as a method of the connection handle
      as well.  For instance, "$db quote $string".

    * Removed unused variables and silence compiler warnings with casting.

    * Added "internal_position" and "internal_query" to the error options.

2007-02-08 Brett Schwarz <brett_schwarz@yahoo.com>
    * Added support for async connections. Note that you will have to 
      use [after] command to find the result of the connection. Callbacks
      won't work, since the sock isn't established yet

2007-01-24 Brett Schwarz <brett_schwarz@yahoo.com>
    * added additional sub commands to dbinfo (version, protocol, param,
    backendpid, socket)

2006-07-05 Karl Lehenbauer <karl-pg@sc.com>
    * Updated autoconf build to Tcl Extension Architecture (TEA) 3.5

2006-01-01 Brett Schwarz <brett_schwarz@yahoo.com>
    * changed the commands that handle params, such that if a value is NULL
      then the lib will change the paramValues value to '\0' so that
      the backend will insert a NULL correctly

2005-11-27 Karl Lehenbauer <karl-pg@sc.com>
    * Bumped version number to 1.6

    * Add pg_null_value_string, ::pg::null_value_string and
      "$conn null_value_string" to get and set a special string value that
      will be returned for fields containing null values from queries.

      pg_select and pg_execute use this value for null values.  If not set,
      they use the traditional empty string.

      pg_exec, pg_exec_prepared, pg_sendquery and pg_sendquery_prepared use
      this value as well.  However, it can be overridden or replaced for
      one specific result using "pg_result $result -null_value_string NULL",
      etc, or by executing the result handle, as in
      "$result -null_value_string NULL"

      Updated documentation and regenerated HTML and PDF files.

2005-11-09 Karl Lehenbauer <karl-pg@sc.com>
    * List -cmdTuples as one of the possible options when executing
      pg_result without arguments.

2005-07-18 Karl Lehenbauer <karl-pg@sc.com>
    * Made pggrok database introspection code much more useful -- routines
      for finding out data about fields and indexes in a table now load
      on array with data about each field (index) and execute a code body
      on it, rather than punting with a call to parray

2005-07-01 Karl Lehenbauer <karl-pg@sc.com>
    * Fixed memory leaks when connections-as-commands are used
      ($conn exec $statement, etc)

2005-04-16 Brett Schwarz <brett_schwarz@yahoo.com>
    * added more documentation about the new capabilities

2005-02-25 Brett Schwarz <brett_schwarz@yahoo.com>
    * took away pg::conninfo and pg::results, and made one command
      pg::dbinfo. Usage is pg::dbinfo connections and 
      pg::dbinfo results $conn

2004-11-30 Karl Lehenbauer <karl-pg@sc.com>
    * Document pg_exec_prepared, pg_results, and pg_sendquery_prepared.

    * Document pg_exec's ability to take optional arguments for 
      $-substitution in PostgreSQL 7.4+.

    * Documented pg_getresult and pg_result -error suboptions.  

    * Pasted connect string options from the libpq manual rather
      than simply referring Pgtcl users to the libpq manual.

    * Regenerated pdf and html documentation.

2004-11-29 Brett Schwarz <brett_schwarz@yahoo.com>
    * Added commands to escape and unescape binary strings

2004-11-29 Karl Lehenbauer <karl-pg@sc.com>
    * Added library of useful PostreSQL helper functions to the playpen.

    * Added dbobjects PostgreSQL/Itcl-object interface package to the playpen.

    * Added pggrok code to introspect into Postgres databases to the playpen.

2004-11-23 Karl Lehenbauer <karl-pg@sc.com>
    * Simplify result object processing

    * Fixed bug in Pg_cancelrequest where the error message would
      not be returned.

    * Remove unused variables.

    * Silence compiler warnings about a fall-through (that couldn't
      be reached) that would cause a function declared to return something
      to not return anything and about a possibly uninitialized variable.

2004-11-22 Karl Lehenbauer <karl-pg@sc.com>
    * Created playpen directory where interesting things that are not
      yet ready for prime time can be examined and messed with.  
      Added code for turning OIDs returned by pg_result -lAttributes
      into data type names.


libpgtcl-1.7/generic/pgtcl.c
/*------------------------------------------------------------------------
 *
 * pgtcl.c
 *
 *	libpgtcl is a tcl package for front-ends to interface with PostgreSQL.
 *	It's a Tcl wrapper for libpq.
 *
 * Portions Copyright (c) 1996-2003, PostgreSQL Global Development Group
 * Portions Copyright (c) 1994, Regents of the University of California
 *
 *
 * IDENTIFICATION
 *	  $Id: pgtcl.c,v 1.31 2008/05/20 15:07:00 karl Exp $
 *
 *-------------------------------------------------------------------------
 */

#include <libpq-fe.h>
#include "libpgtcl.h"
#include "pgtclCmds.h"
#include "pgtclId.h"

#ifdef WIN32
#include <c.h>
#endif

/* BEGIN STUBS MUMBO JUMBO http://mini.net/tcl/1687 */
/* We need at least the Tcl_Obj interface that was started in 8.0 */
#if TCL_MAJOR_VERSION < 8
#error "we need Tcl 8.0 or greater to build this"

/* Check for Stubs compatibility when asked for it. */
#elif defined(USE_TCL_STUBS) && TCL_MAJOR_VERSION == 8 && \
		(TCL_MINOR_VERSION == 0 || \
		(TCL_MINOR_VERSION == 1 && TCL_RELEASE_LEVEL != TCL_FINAL_RELEASE))
#error "Stubs interface doesn't work in 8.0 and alpha/beta 8.1; only 8.1.0+"
#endif

#ifdef _MSC_VER
/* Only do this when MSVC++ is compiling us. */
#ifdef USE_TCL_STUBS
 /* Mark this .obj as needing tcl's Stubs library. */
#pragma comment(lib, "tclstub" \
			STRINGIFY(JOIN(TCL_MAJOR_VERSION,TCL_MINOR_VERSION)) ".lib")
#if !defined(_MT) || !defined(_DLL) || defined(_DEBUG)

 /*
  * This fixes a bug with how the Stubs library was compiled. The
  * requirement for msvcrt.lib from tclstubXX.lib should be removed.
  */
#pragma comment(linker, "-nodefaultlib:msvcrt.lib")
#endif
#else
 /* Mark this .obj needing the import library */
#pragma comment(lib, "tcl" \
		STRINGIFY(JOIN(TCL_MAJOR_VERSION,TCL_MINOR_VERSION)) ".lib")
#endif
#endif
#undef TCL_STORAGE_CLASS
#define TCL_STORAGE_CLASS DLLEXPORT
/* END STUBS MUMBO JUMBO */



typedef struct {
    char *name;                 /* Name of command. */
    char *name2;                /* Name of command, in ::pg namespace. */
    Tcl_ObjCmdProc *objProc;    /* Command's object-based procedure. */
    int protocol;    /* version 2 or version 3 (>=7.4) of PG protocol */
} PgCmd;

static PgCmd commands[] = {
    {"pg_conndefaults", "::pg::conndefaults", Pg_conndefaults, 2},
    {"pg_connect", "::pg::connect", Pg_connect,2},
    {"pg_disconnect", "::pg::disconnect", Pg_disconnect,2},
    {"pg_exec", "::pg::sqlexec", Pg_exec,2},
    {"pg_exec_prepared", "::pg::exec_prepared", Pg_exec_prepared,3},
    {"pg_select", "::pg::select", Pg_select,2},
    {"pg_result", "::pg::result", Pg_result,2},
    {"pg_execute", "::pg::execute", Pg_execute,2},
    {"pg_lo_open", "::pg::lo_open", Pg_lo_open,2},
    {"pg_lo_close", "::pg::lo_close", Pg_lo_close,2},
    {"pg_lo_read", "::pg::lo_read", Pg_lo_read,2},
    {"pg_lo_write", "::pg::lo_write", Pg_lo_write,2},
    {"pg_lo_lseek", "::pg::lo_lseek", Pg_lo_lseek,2},
    {"pg_lo_creat", "::pg::lo_creat", Pg_lo_creat,2},
    {"pg_lo_tell", "::pg::lo_tell", Pg_lo_tell,2},
    {"pg_lo_truncate", "::pg::lo_truncate", Pg_lo_truncate,2},
    {"pg_lo_unlink", "::pg::lo_unlink", Pg_lo_unlink,2},
    {"pg_lo_import", "::pg::lo_import", Pg_lo_import,2},
    {"pg_lo_export", "::pg::lo_export", Pg_lo_export,2},
    {"pg_listen", "::pg::listen", Pg_listen,2},
    {"pg_sendquery", "::pg::sendquery", Pg_sendquery,2},
    {"pg_sendquery_prepared", "::pg::sendquery_prepared", Pg_sendquery_prepared,3},
    {"pg_getresult", "::pg::getresult", Pg_getresult,2},
    {"pg_isbusy", "::pg::isbusy", Pg_isbusy,2},
    {"pg_blocking", "::pg::blocking", Pg_blocking,2},
    {"pg_null_value_string", "::pg::null_value_string", Pg_null_value_string,2},
    {"pg_cancelrequest", "::pg::cancelrequest", Pg_cancelrequest,2},
    {"pg_on_connection_loss", "::pg::on_connection_loss", Pg_on_connection_loss,2},
    {"pg_quote", "::pg::quote", Pg_quote,2},
    {"pg_escape_string", "::pg::escape_string", Pg_quote,2},
    {"pg_escape_bytea", "::pg::escape_bytea", Pg_escapeBytea,2},
    {"pg_unescape_bytea", "::pg::unescape_bytea", Pg_unescapeBytea,2},
    {"pg_dbinfo", "::pg::dbinfo", Pg_dbinfo,2},
    {"pg_getdata", "::pg::getdata", Pg_getdata,2},
    {NULL, NULL, NULL, 0}
};



/*
 * Pgtcl_Init
 *	  initialization package for the PGTCL Tcl package
 *
 */

EXTERN int
Pgtcl_Init(Tcl_Interp *interp)
{
	double		tclversion;
	Tcl_Obj    *tclVersionObj;
        PgCmd *cmdPtr;

        #ifdef WIN32
        WSADATA wsaData;
        #endif

#ifdef USE_TCL_STUBS
	if (Tcl_InitStubs(interp, "8.1", 0) == NULL)
		return TCL_ERROR;
#endif

        #ifdef WIN32X
        /*
        * On Windows, need to explicitly load the libpq library to
        * force the call to WSAStartup.
        */
	Tcl_Obj    *tresult;

        if (LoadLibrary("libpq.dll") == NULL) {
        //char buf[32];
        //sprintf(buf, "%d", GetLastError());
        tresult = Tcl_NewStringObj("Cannot load \"libpq.dll\" (or dependant), error was ");
        Tcl_AppendToObj(tresult, GetLastError(), -1);
        Tcl_SetObjResult(interp, tresult);
        
/*
        Tcl_AppendStringsToObj(Tcl_GetObjResult(interp), 
            "Cannot load \"libpq.dll\" (or dependant), error was ", 
            GetLastError(), NULL);
*/

        return TCL_ERROR;
        }
        #endif
        
    #ifdef WIN32

            if (WSAStartup(MAKEWORD(1, 1), &wsaData))
            {
                   /*
                    * No really good way to do error handling here, since we
                    * don't know how we were loaded
                    */
                    return FALSE;
            }

    #endif


	/*
	 * Tcl versions >= 8.1 use UTF-8 for their internal string
	 * representation. Therefore PGCLIENTENCODING must be set to UNICODE
	 * for these versions.
	 */

	if ((tclVersionObj = Tcl_GetVar2Ex(interp, "tcl_version", NULL, TCL_GLOBAL_ONLY)) == NULL)
		return TCL_ERROR;

	if (Tcl_GetDoubleFromObj(interp, tclVersionObj, &tclversion) == TCL_ERROR)
		return TCL_ERROR;

	if (tclversion >= 8.1)
		Tcl_PutEnv("PGCLIENTENCODING=UNICODE");

	/* register all pgtcl commands */


    for (cmdPtr = commands; cmdPtr->name != NULL; cmdPtr++) {

        Tcl_CreateObjCommand(interp, cmdPtr->name, 
             cmdPtr->objProc, (ClientData) "::",NULL);
         Tcl_CreateObjCommand(interp, cmdPtr->name2, 
             cmdPtr->objProc, (ClientData) "::pg::",NULL);
    }


    if (Tcl_Eval(interp, "namespace eval ::pg namespace export *") == TCL_ERROR)
        return TCL_ERROR;


	return Tcl_PkgProvide(interp, "Pgtcl", PACKAGE_VERSION);
}

int
Pgtcl_SafeInit(Tcl_Interp *interp)
{
	return Pgtcl_Init(interp);
}
